File: frontend/src/services/testPlan.ts
import api, { type ApiResponse, type PaginatedResponse } from './api';

// 类型定义
export interface TestPlan {
  id: number;
  name: string;
  description?: string;
  execution_strategy: 'sequential' | 'parallel';
  max_parallel_tasks: number;
  enabled: boolean;
  project_id?: number;
  suite_count: number;
  created_at: string;
  updated_at: string;
}

export interface TestPlanDetail extends TestPlan {
  suites: TestPlanSuiteDetail[];
}

export interface TestPlanSuite {
  suite_id: number;
  suite_name?: string;
  order: number;
  enabled: boolean;
  execution_config?: Record<string, unknown>;
}

export interface TestPlanTestcaseOrder {
  testcase_id: number;
  testcase_name?: string;
  order_index: number;
  priority?: string;
}

export interface TestPlanSuiteDetail {
  id: number;
  test_plan_id: number;
  suite_id: number;
  suite_name?: string;
  order_index: number;
  enabled: boolean;
  testcases: TestPlanTestcaseOrder[];
}

export interface TestPlanCreate {
  name: string;
  description?: string;
  execution_strategy?: 'sequential' | 'parallel';
  max_parallel_tasks?: number;
  enabled?: boolean;
  project_id?: number;
  suites?: TestPlanSuite[];
}

export interface TestPlanUpdate {
  name?: string;
  description?: string;
  execution_strategy?: 'sequential' | 'parallel';
  max_parallel_tasks?: number;
  enabled?: boolean;
  suites?: TestPlanSuite[];
}

export interface TestPlanExecuteRequest {
  platform: string;
  device_serial?: string;
  profile_id?: number;
  timeout?: number;
  extra_args?: Record<string, unknown>;
  priority?: string;
}

// 获取测试计划列表
export const getTestPlans = (params?: {
  page?: number;
  page_size?: number;
  project_id?: number;
}) => {
  return api.get<ApiResponse<PaginatedResponse<TestPlan>>>('/test-plans', { params });
};

// 获取测试计划详情
export const getTestPlan = (id: number) => {
  return api.get<ApiResponse<TestPlanDetail>>(`/test-plans/${id}`);
};

// 创建测试计划
export const createTestPlan = (data: TestPlanCreate) => {
  return api.post<ApiResponse<TestPlan>>('/test-plans', data);
};

// 更新测试计划
export const updateTestPlan = (id: number, data: TestPlanUpdate) => {
  return api.put<ApiResponse<TestPlan>>(`/test-plans/${id}`, data);
};

// 删除测试计划
export const deleteTestPlan = (id: number) => {
  return api.delete<ApiResponse>(`/test-plans/${id}`);
};

// 切换测试计划启用状态
export const toggleTestPlan = (id: number, enabled: boolean) => {
  return api.patch<ApiResponse<TestPlan>>(`/test-plans/${id}/toggle`, { enabled });
};

// 添加套件到测试计划
export const addSuitesToPlan = (id: number, suite_ids: number[]) => {
  return api.post<ApiResponse<TestPlanDetail>>(`/test-plans/${id}/suites`, { suite_ids });
};

// 从测试计划移除套件
export const removeSuitesFromPlan = (id: number, suite_ids: number[]) => {
  return api.delete<ApiResponse<TestPlanDetail>>(`/test-plans/${id}/suites`, { data: { suite_ids } });
};

// 调整套件顺序
export const reorderPlanSuites = (id: number, suites: TestPlanSuite[]) => {
  return api.put<ApiResponse<TestPlanDetail>>(`/test-plans/${id}/suites/reorder`, { suites });
};

// 设置用例顺序
export const setSuiteTestcaseOrder = (
  planId: number,
  suiteId: number,
  testcase_orders: TestPlanTestcaseOrder[]
) => {
  return api.put<ApiResponse<TestPlanDetail>>(
    `/test-plans/${planId}/suites/${suiteId}/testcases/order`,
    { testcase_orders }
  );
};

// 执行测试计划
export const executeTestPlan = (id: number, data: TestPlanExecuteRequest) => {
  return api.post<ApiResponse<RunResponse[]>>(`/test-plans/${id}/run`, data);
};

// 执行响应类型
export interface RunResponse {
  task_id: string;
  type: string;
  targets: Array<{ id: number; name: string }>;
  status: string;
  cmd: string[];
  started_at: string;
}

File: backend/app/schemas/test_plan.py
"""
TestPlan schemas
"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from datetime import datetime


class TestPlanSuiteSchema(BaseModel):
    """Test plan suite association"""
    suite_id: int = Field(..., description="Suite ID")
    suite_name: Optional[str] = Field(None, description="Suite name")
    order: int = Field(..., ge=1, description="Execution order in the plan")
    enabled: bool = Field(True, description="Enabled flag")
    execution_config: Optional[Dict[str, Any]] = Field(None, description="Execution configuration")


class TestPlanTestcaseOrderSchema(BaseModel):
    """Testcase order in test plan suite"""
    testcase_id: int = Field(..., description="Testcase ID")
    testcase_name: Optional[str] = Field(None, description="Testcase name")
    order_index: int = Field(..., ge=0, description="Execution order")
    priority: Optional[str] = Field(None, description="Testcase priority")

    class Config:
        from_attributes = True


class TestPlanBase(BaseModel):
    """Base test plan fields"""
    name: str = Field(..., min_length=1, max_length=200, description="Test plan name")
    description: Optional[str] = Field(None, description="Description")
    execution_strategy: str = Field("sequential", pattern="^(sequential|parallel)$", description="Execution strategy")
    max_parallel_tasks: int = Field(1, ge=1, le=10, description="Max parallel tasks")
    enabled: bool = Field(True, description="Enabled flag")


class TestPlanCreate(TestPlanBase):
    """Test plan creation schema"""
    project_id: Optional[int] = Field(None, description="Project ID")
    suites: Optional[List[TestPlanSuiteSchema]] = Field(None, min_length=1, description="List of suites")


class TestPlanUpdate(BaseModel):
    """Test plan update schema"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    execution_strategy: Optional[str] = Field(None, pattern="^(sequential|parallel)$")
    max_parallel_tasks: Optional[int] = Field(None, ge=1, le=10)
    enabled: Optional[bool] = None
    suites: Optional[List[TestPlanSuiteSchema]] = None


class TestPlanSuiteDetailSchema(BaseModel):
    """Test plan suite with testcases"""
    id: int = Field(..., description="Test plan suite ID")
    test_plan_id: int = Field(..., description="Test plan ID")
    suite_id: int = Field(..., description="Suite ID")
    suite_name: Optional[str] = Field(None, description="Suite name")
    order_index: int = Field(..., description="Execution order")
    enabled: bool = Field(..., description="Enabled flag")
    testcases: List[TestPlanTestcaseOrderSchema] = Field(default_factory=list, description="Testcases in order")


class TestPlanResponse(TestPlanBase):
    """Test plan response schema"""
    id: int
    project_id: Optional[int] = None
    suite_count: int = Field(0, description="Number of suites")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TestPlanDetailResponse(TestPlanResponse):
    """Test plan detail with suites"""
    suites: List[TestPlanSuiteDetailSchema] = Field(default_factory=list, description="Suites in the plan")


class TestPlanToggleRequest(BaseModel):
    """Test plan toggle request"""
    enabled: bool = Field(..., description="Enabled state")


class TestPlanAddSuitesRequest(BaseModel):
    """Add suites to test plan"""
    suite_ids: List[int] = Field(..., min_length=1, description="Suite IDs to add")


class TestPlanRemoveSuitesRequest(BaseModel):
    """Remove suites from test plan"""
    suite_ids: List[int] = Field(..., min_length=1, description="Suite IDs to remove")


class TestPlanReorderSuitesRequest(BaseModel):
    """Reorder suites in test plan"""
    suites: List[TestPlanSuiteSchema] = Field(..., min_length=1, description="Suites with new order")


class TestPlanSetTestcaseOrderRequest(BaseModel):
    """Set testcase order for a suite"""
    testcase_orders: List[TestPlanTestcaseOrderSchema] = Field(..., min_length=1, description="Testcase orders")


class TestPlanExecuteRequest(BaseModel):
    """Execute test plan request"""
    platform: str = Field(..., description="Execution platform (android/ios/web)")
    device_serial: Optional[str] = Field(None, description="Device serial for execution")
    profile_id: Optional[int] = Field(None, description="Profile ID for execution")
    timeout: Optional[int] = Field(None, ge=1, description="Execution timeout in seconds")
    extra_args: Optional[Dict[str, Any]] = Field(None, description="Extra execution arguments")
    priority: Optional[str] = Field("normal", description="Execution priority")


class TestPlanExecuteResponse(BaseModel):
    """Test plan execution response"""
    task_id: str = Field(..., description="Execution task ID")
    runs_count: int = Field(..., description="Number of runs created")
